### PR TITLE
Migrate core interceptors to use InterceptorType CRD

### LINKED ISSUE #868: Interceptors: Implement CRD to provide a single way to install Interceptors
The work for this will involve implementing the InterceptorConfig CRD as described in [TEP-0026](https://github.com/tektoncd/community/blob/master/teps/0026-interceptor-plugins.md#installing-interceptors)

### PR DESCRIPTION
**Pull request by dibyom**
# Changes

Migrate core interceptors to use InterceptorType CRD

This commit does the following:
1. Add YAML definitions for installing each core interceptor using the
new InterceptorType CRD.
2. Modify the EventListener to use the InterceptorType CRD to find the
Interceptor's URL.

Fixes #868

This is NOT a breaking change since users can still configure their interceptors in the same way was before. Changes to the way a TriggerInterceptor is configured is coming in #869.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#tests) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#docs) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commits)
- [ ] Release notes block has been filled in or deleted (only if no user facing changes)

_See [the contribution guide](https://github.com/tektoncd/triggers/blob/master/CONTRIBUTING.md) for more details._

# Release Notes

Depends on #960

### CODE REVIEW

File: pkg/interceptors/interceptors_test.go [hunk 1]
```diff
@@ -240,7 +243,7 @@ func TestGetInterceptorParams_Error(t *testing.T) {
 	}
 }
 
-func Test_GetSecretToken(t *testing.T) {
+func TestGetSecretToken(t *testing.T) {
 	tests := []struct {
 		name   string
 		cache  map[string]interface{}
```
Comment (bigkevmcd) [inline pkg/interceptors/interceptors_test.go:249]: It's a bit confusing from the diff, but I suspect that `TestResolvePath` which is the function for the changes below should be renamed to `TestGetName` ?

File: pkg/reconciler/v1alpha1/clusterinterceptor/controller.go [hunk 2]
```diff
@@ -27,11 +27,11 @@ import (
 	"knative.dev/pkg/logging"
 )
 
-// NewController creates a new instance of an InterceptorType controller.
+// NewController creates a new instance of an ClusterInterceptor controller.
 func NewController() func(context.Context, configmap.Watcher) *controller.Impl {
 	return func(ctx context.Context, cmw configmap.Watcher) *controller.Impl {
 		logger := logging.FromContext(ctx)
-		interceptorTypeInformer := clusterinterceptorinformer.Get(ctx)
+		ClusterInterceptorInformer := clusterinterceptorinformer.Get(ctx)
```
Comment (bigkevmcd) [inline pkg/reconciler/v1alpha1/clusterinterceptor/controller.go:34]: Is there a reason to make this local look like an exported value?

File: pkg/interceptors/interceptors_test.go [hunk 3]
```diff
@@ -305,48 +308,106 @@ func TestResolvePath(t *testing.T) {
...
+	}}
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			got, err := interceptors.ResolveToURL(tc.getter, tc.itype)
+			if err != nil {
+				t.Fatalf("ResolveToURL() error: %s", err)
+			}
+
+			if got.String() != tc.want {
+				t.Fatalf("ResolveToURL want: %s; got: %s", tc.want, got)
```
Comment (bigkevmcd) [inline pkg/interceptors/interceptors_test.go:393]: I'd probably prefer this to be consistent `ResolveToURL()`

File: pkg/interceptors/interceptors.go [hunk 4]
```diff
@@ -171,28 +170,40 @@ func UnmarshalParams(ip map[string]interface{}, p interface{}) error {
...
+	return name
+}
+
+type InterceptorGetter func(name string) (*triggersv1.ClusterInterceptor, error)
+
+// ResolveToURL finds an Interceptor's URL.
+func ResolveToURL(getter InterceptorGetter, name string) (*apis.URL, error) {
+	ic, err := getter(name)
+	if err != nil {
+		return nil, err
```
Comment (bigkevmcd) [inline pkg/interceptors/interceptors.go:196]: I think we can improve the error message here to indicate what it was trying to find?

**Review by bigkevmcd — COMMENTED** [review]
Some relatively minor mostly cosmetic issues, otherwise looks good to me.

File: pkg/interceptors/interceptors.go [hunk 5]
```diff
@@ -171,28 +170,40 @@ func UnmarshalParams(ip map[string]interface{}, p interface{}) error {
 	return nil
 }
 
-// ResolveURL returns the URL for the given core interceptor
-func ResolveURL(i *triggersv1.TriggerInterceptor) *url.URL {
+// Getname returns the name for the given core interceptor
+func GetName(i *triggersv1.TriggerInterceptor) string {
```
[the remark this answers is not in the document]
Comment (dibyom) [reply pkg/interceptors/interceptors.go:174]: Done.

File: pkg/interceptors/interceptors_test.go (hunk 1 above)
Comment (dibyom) [reply pkg/interceptors/interceptors_test.go:249]: yup good catch, changed the name and made it part of TriggerInterceptor.

File: pkg/reconciler/v1alpha1/clusterinterceptor/controller.go (hunk 2 above)
Comment (dibyom) [reply pkg/reconciler/v1alpha1/clusterinterceptor/controller.go:34]: I think this was due to a massive search and replace I did

**Review by savitaashture — COMMENTED** [review]
/lgtm

Verified locally it works as expected
Thank you :+1: 

Just one question if user want to write custom interceptor what all things they have to take care? do we have any steps/instructions?

**Comment by dibyom** [comment]
> Just one question if user want to write custom interceptor what all things they have to take care? do we have any steps/instructions?

Docs coming for those as part of https://github.com/tektoncd/triggers/pull/1001

**Review by savitaashture — APPROVED** [review]

**Pull request merged**